src/IO/Importers/ElementActions.hpp

```diff
@@ -53,5 +61,111 @@ struct RegisterWithVolumeDataReader {
   }
 };
 
-}  // namespace Actions
-}  // namespace importers
+/*!
+ * \brief Read a volume data file and distribute the data to all registered
+ * elements.
+ *
+ * Invoke this action on the elements of an array parallel component to dispatch
+ * reading the volume data file specified by the options in
+ * `ImporterOptionsGroup`. The tensors in `FieldTagsList` will be loaded from
+ * the file and distributed to all elements that have previously registered. Use
+ * `importers::Actions::RegisterWithVolumeDataReader` to register the elements
+ * of the array parallel component in a previous phase.
+ *
+ * Note that the volume data file will only be read once per node, triggered by
+ * the first element that invokes this action. All subsequent invocations of
+ * this action on the node will do nothing. See
+ * `importers::Actions::ReadAllVolumeDataAndDistribute` for details.
+ *
+ * The data is distributed to the elements using `Parallel::receive_data`. The
+ * elements can monitor `importers::Tags::VolumeData` in their inbox to wait for
+ * the data and process it once it's available. We provide the action
+ * `importers::Actions::ReceiveVolumeData` that waits for the data and moves it
+ * directly into the DataBox. You can also implement a specialized action that
+ * might verify and post-process the data before populating the DataBox.
+ *
+ * \see \ref dev_guide_importing
+ */
+template <typename ImporterOptionsGroup, typename FieldTagsList>
+struct ReadVolumeData {
+  using const_global_cache_tags =
+      tmpl::list<Tags::FileName<ImporterOptionsGroup>,
+                 Tags::Subgroup<ImporterOptionsGroup>,
+                 Tags::ObservationValue<ImporterOptionsGroup>>;
+
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&> apply(
+      db::DataBox<DbTagsList>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      Parallel::GlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    auto& local_reader_component =
+        *Parallel::get_parallel_component<
+             importers::VolumeDataReader<Metavariables>>(cache)
+             .ckLocalBranch();
```

Comment:
I half remember in our discussion yesterday that we should avoid calling .ckLocalBranch? So I understand why it's ok, could you please explain how this line works and remind me what the concern about .ckLocalBranch was?

Reply:
The `VolumeDataReader` nodegroup component has one chare on every node, and here I just retrieve that chare directly.

Reply:
While @wthrowe is right, the `receive_data` calls may be automatically inlined and so are synchronous. This is not true for `simple_action`, and so to be safe we need to make this an async call through Charm++

Reply:
Is it safe to call `receive_data` from a nodegroup at all then?  The asynchronous nodegroup action could be scheduled by Charm++ to run at the same time as an element action on a different thread, couldn't it?

Reply:
Sorry, I'm not sure I exactly understand the scenario you mean. If I understand correctly, you're worried about the nodegroup calling `receive_data` while the element is being worked on on another core? So, say the nodegroup action is being run on core 2, and it invokes `receive_data` on a element that is on core 4 (same node, different core). This call, as far as I understand Charm++, would not be inlined because `core 4` is treated as non-local for array chares, but local for nodegroups. That is, it is the callee that determines whether or not it is local, not the caller, I think.

If I've completely misunderstood the discussion, please let me know :)

Reply:
Ah, I see. That makes sense.

Reply:
I removed the `ckLocalBranch` in a fixup commit